lib/OpenQA/Task/Needle/Save.pm

```diff
@@ -78,21 +81,26 @@ sub _save_needle {
         return $minion_job->finish({error => "<strong>Failed to validate $needlename.</strong><br>$error"});
     }
 
-    # determine imagepath
+    my @match_areas = grep { $_->{type} eq 'match' } @{$json_data->{area}};
+    my $no_match_areas = not scalar(@match_areas);
+
     my $imagepath;
-    if ($imagedir) {
-        $imagepath = join('/', $imagedir, $imagename);
-    }
-    elsif ($imagedistri) {
-        $imagepath = join('/', needledir($imagedistri, $imageversion), $imagename);
-    }
-    else {
-        $imagepath = join('/', $openqa_job->result_dir(), $imagename);
-    }
-    if (!-f $imagepath) {
-        my $error = "Image $imagename could not be found!";
-        $app->log->error("Failed to save needle: $error");
-        return $minion_job->fail({error => "<strong>Failed to save $needlename.</strong><br>$error"});
+    unless ($no_match_areas) {
```

Comment:
The double-negative is not very nice and we won't need to extra variable here.

```suggestion
    if (@match_areas) {
```

The same counts for the other `unless ($no_match_areas)` occurrences.